client/src/main/java/io/pravega/client/stream/impl/CheckpointState.java

```diff
@@ -159,5 +173,59 @@ public String toString() {
         sb.append(uncheckpointedHosts.toString());
         sb.append(" }");
         return sb.toString();
-    }    
+    }
+
+    @VisibleForTesting
+    static class CheckpointStateBuilder implements ObjectBuilder<CheckpointState> {
+
+    }
+
+    private static class CheckpointStateSerializer
+            extends VersionedSerializer.WithBuilder<CheckpointState, CheckpointStateBuilder> {
+        @Override
+        protected CheckpointStateBuilder newBuilder() {
+            return builder();
+        }
+
+        @Override
+        protected byte getWriteVersion() {
+            return 0;
+        }
+
+        @Override
+        protected void declareVersions() {
+            version(0).revision(0, this::write00, this::read00);
+        }
+        
+        private void read00(RevisionDataInput input, CheckpointStateBuilder builder) throws IOException {
+            ElementDeserializer<String> stringDeserializer = RevisionDataInput::readUTF;
+            ElementDeserializer<Long> longDeserializer = RevisionDataInput::readLong;
+            ElementDeserializer<Segment> segmentDeserializer = in -> Segment.fromScopedName(in.readUTF());
+            builder.checkpoints(new ArrayList<>(input.readCollection(stringDeserializer)));
```

Comment:
pass `ArrayList::new` to `readCollection` and it will create a new ArrayList for you; no need to double-copy.